### PR TITLE
SHIP-0004: Build Environment Variables 

### PR DESCRIPTION
**Pull request by coreydaley**
# Changes

# Submitter Checklist

- [x] Includes tests if functionality changed/was added
- [x] Includes docs if changes are user-facing
- [x] [Set a kind label on this PR](https://prow.k8s.io/command-help#kind)  
- [x] Release notes block has been filled in, or marked NONE

See [the contributor guide](https://github.com/shipwright-io/build/blob/main/CONTRIBUTING.md)
for details on coding conventions, github and prow interactions, and the code review process.

# Release Notes
Adds the ability to specify environment variables for `Builds` and `BuildRuns` as a `key=value` pair via the command line using the `--env` flag.
Environment variables specified for a `BuildRun` will override those in the `Build`.
```
Example:  
    $ shp build create my-build --env VAR_1=value-1  
    $ shp buildrun create my-buildrun --env VAR_2=value-2
```

### CODE REVIEW

**Comment by coreydaley** [comment]
/hold
depends on https://github.com/shipwright-io/build/pull/817 and removing line from go.mod

**Comment by gabemontero** [comment]
> # Changes
> # Submitter Checklist
> 
>     * [x]  Includes tests if functionality changed/was added
> 
>     * [ ]  Includes docs if changes are user-facing
> 
>     * [ ]  [Set a kind label on this PR](https://prow.k8s.io/command-help#kind)
> 
>     * [ ]  Release notes block has been filled in, or marked NONE
> 
> 
> See [the contributor guide](https://github.com/shipwright-io/build/blob/main/CONTRIBUTING.md) for details on coding conventions, github and prow interactions, and the code review process.
> # Release Notes

release note formattting is not quite right ... nothing is rendoring underneath the heading as I currently view it

**Comment by gabemontero** [comment]
otherwise, since I see no one assigned, am I correct @coreydaley that in addition to the hold you have here, this is still a work in progress item, and we should hold off on reviewing?

Or should I add this to the queue and dive in?

thanks

**Comment by coreydaley** [comment]
@gabemontero 
 - I have updated the Release Notes section.
 - Yes, you can go ahead and do a review, keeping in mind the comments that you left on shipwright-io/build and that we are waiting on @adambkaplan to make a decision in regards to the afformentioned.

**Comment by coreydaley** [comment]
These tests won't pass until https://github.com/shipwright-io/build/pull/817 merges

File: test/e2e/e2e.bats
Comment (gabemontero) [inline test/e2e/e2e.bats:26]: @otaviof @coreydaley - I have not tried it yet, but perhaps you all have.  With the kind/local registry stuff we are employing, is it correct to say that by default we do *NOT* need credentials to push images to it?

And hence why I don't see cred creation here?

**Review by gabemontero — COMMENTED** [review]
Yeah pending the type for the env var array per my comments in your associated build PR nothing immediate that I see @coreydaley 

I did ask a question around the kind / local registry thing ... maybe a change needed there based on the answer (but I'm guessing no)

and lots of useful libs around BATS and the e2e's .... I plan to wait until all that merges so I can build off of it for adding an e2e for `ship build run <name> --follow` :-)

**Comment by gabemontero** [comment]
/hold

just in case, as we need to vendor in the associated build PR

**Comment by gabemontero** [comment]
> @gabemontero
> 
>     * I have updated the Release Notes section.
> 
>     * Yes, you can go ahead and do a review, keeping in mind the comments that you left on shipwright-io/build and that we are waiting on @adambkaplan to make a decision in regards to the afformentioned.

some cross referencing with  https://github.com/shipwright-io/build/pull/817#issuecomment-930221460 and https://github.com/shipwright-io/build/pull/817#issuecomment-930228329

summary
- @coreydaley thinks he'll be able to add all the k8s corev1.envvar *valuefrom related doc and tests, so his change will continue to use that k8s type
- most likely this cli pr will not take on creating env vars from those supplied sources ... we'll track that as a potential add on based on user feedback, with the history that with the analogous support in openshift builds, we never took on cli level support for adding valuefrom env vars

File: test/e2e/e2e.bats
Comment (otaviof) [reply test/e2e/e2e.bats:26]: Yes, we can assume the registry won't have any authentication, since we are in complete control of the container registry in use after #43 👍🏼

File: test/e2e/e2e.bats
Comment (otaviof) [inline test/e2e/e2e.bats:6]: As much as I like the syntax and the practical effects of the "bats-support" helper functions, I don't like the idea of copying the whole project. I think we could create our own helper functions and make use local tools like `grep`, `awk` and others.

File: pkg/shp/cmd/buildrun/create.go
Comment (otaviof) [inline pkg/shp/cmd/buildrun/create.go:66]: Why do we need to copy it back to the `BuildRun` instance? I think it should be using the `BuildRun` instance wired by `flags` package, thus it should have the `env` slice already populated.

File: pkg/shp/cmd/buildrun/create.go
Comment (otaviof) [inline pkg/shp/cmd/buildrun/create.go:70]: This type of action should be done during `Complete()` (or maybe `Prepare()`), so the `Run` method can concentrate on executing the task.

File: pkg/shp/flags/build.go
Comment (otaviof) [inline pkg/shp/flags/build.go:18]: What's the reason for this change?

**Review by otaviof — CHANGES_REQUESTED** [review]
Some comments left on the pull-request, and I'm having issues with the Go modules changed. As per:

```shell
$ go mod vendor
go: github.com/shipwright-io/build@v0.5.2-0.20210715083206-5d8fb411a1eb (replaced by github.com/coreydaley/shipwright-io-build@v0.0.0-20210927153915-3175d47a83e4): version "v0.0.0-20210927153915-3175d47a83e4" invalid: unknown revision 3175d47a83e4
go: downloading github.com/coreydaley/shipwright-io-build v0.0.0-20210927153915-3175d47a83e4
go: downloading github.com/spf13/cobra v1.2.1
go: downloading k8s.io/apimachinery v0.20.6
go: downloading k8s.io/client-go v0.20.2
go: downloading k8s.io/api v0.20.6
go: downloading github.com/onsi/gomega v1.10.3
go: downloading k8s.io/cli-runtime v0.20.2
go: downloading github.com/texttheater/golang-levenshtein/levenshtein v0.0.0-20200805054039-cae8b0eaed6c
go: downloading k8s.io/utils v0.0.0-20210111153108-fddb29f9d009
go: github.com/shipwright-io/build@v0.5.2-0.20210715083206-5d8fb411a1eb (replaced by github.com/coreydaley/shipwright-io-build@v0.0.0-20210927153915-3175d47a83e4): version "v0.0.0-20210927153915-3175d47a83e4" invalid: unknown revision 3175d47a83e4
```
Same with `go mod tidy`, for the record.

File: test/e2e/e2e.bats
Comment (coreydaley) [reply test/e2e/e2e.bats:6]: That seems a bit silly to basically re-write a library that is provided by the framework that we want to use.
I would be open to adding it as a submodule.

File: pkg/shp/cmd/buildrun/create.go
Comment (coreydaley) [reply pkg/shp/cmd/buildrun/create.go:66]: The flag can not have a type of corev1.EnvVar, so the array of strings needs to be processed into that type.

File: pkg/shp/cmd/buildrun/create.go
Comment (coreydaley) [reply pkg/shp/cmd/buildrun/create.go:70]: The BuildRun object does not exist until the `Run` function ....

File: pkg/shp/flags/build.go
Comment (coreydaley) [reply pkg/shp/flags/build.go:18]: I think there was an issue when one of the dependencies was updated.

**Comment by coreydaley** [comment]
> Some comments left on the pull-request, and I'm having issues with the Go modules changed. As per:
> 
> ```shell
> $ go mod vendor
> go: github.com/shipwright-io/build@v0.5.2-0.20210715083206-5d8fb411a1eb (replaced by github.com/coreydaley/shipwright-io-build@v0.0.0-20210927153915-3175d47a83e4): version "v0.0.0-20210927153915-3175d47a83e4" invalid: unknown revision 3175d47a83e4
> go: downloading github.com/coreydaley/shipwright-io-build v0.0.0-20210927153915-3175d47a83e4
> go: downloading github.com/spf13/cobra v1.2.1
> go: downloading k8s.io/apimachinery v0.20.6
> go: downloading k8s.io/client-go v0.20.2
> go: downloading k8s.io/api v0.20.6
> go: downloading github.com/onsi/gomega v1.10.3
> go: downloading k8s.io/cli-runtime v0.20.2
> go: downloading github.com/texttheater/golang-levenshtein/levenshtein v0.0.0-20200805054039-cae8b0eaed6c
> go: downloading k8s.io/utils v0.0.0-20210111153108-fddb29f9d009
> go: github.com/shipwright-io/build@v0.5.2-0.20210715083206-5d8fb411a1eb (replaced by github.com/coreydaley/shipwright-io-build@v0.0.0-20210927153915-3175d47a83e4): version "v0.0.0-20210927153915-3175d47a83e4" invalid: unknown revision 3175d47a83e4
> ```
> 
> Same with `go mod tidy`, for the record.

Correct, since I am pulling in source from my own branch until my other api changes merge in shipwright-io/build.

File: test/e2e/e2e.bats
Comment (gabemontero) [reply test/e2e/e2e.bats:6]: hmmm ... I see both sides here

don't know the answer to this question:  even though bats is not a go project, can we use go mod to vendor the repo in, and then load from the vendored location?

Comment (coreydaley) [reply test/e2e/e2e.bats:6]: That seems a bit sketchy to me, I think I would rather the submodule.

Comment (gabemontero) [reply test/e2e/e2e.bats:6]: sorry multitasking too much this AM ... can you clarify / further explain what you mean by submodule ?

Comment (coreydaley) [reply test/e2e/e2e.bats:6]: using a git submodule to pull in those projects.

Comment (gabemontero) [reply test/e2e/e2e.bats:6]: ah *those* submodules :-)

yes that feels like the appropriate solution to achieve the same end goal as the go mod vendor idea ... just been forever since I even thought about git submodules

so yeah since you are open to the idea @coreydaley please do refactor this PR to pull in the bats-core helpers in that fashion, assuming using git submodules in our project have not blockers .... I have no idea if this is the case, but I would not be surprised if @adambkaplan as someone with more administrative powers to add the submodule to the "parent" project i.e. our cli repo here

Comment (coreydaley) [reply test/e2e/e2e.bats]: code has been updated to pull in BATS via git submodules and run without having to install BATS locally. ptal
Note that tests won't pass until the shipwright-io/build pr merges

Comment (otaviof) [reply test/e2e/e2e.bats]: > hmmm ... I see both sides here
> 
> don't know the answer to this question: even though bats is not a go project, can we use go mod to vendor the repo in, and then load from the vendored location?

I like this idea! Would be easier to deal with than "git sub-modules" IMO, and it does solve the issue of keeping track of another external repository. 👍🏼

File: pkg/shp/cmd/buildrun/create.go
Comment (otaviof) [reply pkg/shp/cmd/buildrun/create.go:70]: Maybe not as a subject for this pull-request, but being it the `buildrun` sub-command, I think it would make sense to deal with a `BuildRun` object instead of a `Build`... thinking out loud.

File: test/e2e/e2e.bats
Comment (gabemontero) [reply test/e2e/e2e.bats]: looking at @coreydaley 's https://github.com/shipwright-io/cli/pull/27/commits/2150530e2c142c4bc2b7c9f791783324931a13f8 I actually agree with him @otaviof that we should use git submodules for this ... I agree go mod vendoring should not be used for a project like BATS which is not go based

File: pkg/shp/cmd/buildrun/create.go
Comment (gabemontero) [reply pkg/shp/cmd/buildrun/create.go:70]: as I recall env vars in the shipwright-io/build PR and associated SHIP EP can be set on both the Build and BuildRun

reasonable defaults are set on the Build

overrides on atypical adds (i.e. not set on the Build) are set on the BuildRun

discussion on the pros and cons of this, if you want to continue @otaviof , as you suggested, should span beyond this PR

given where the EP and build PR is, I would say we move forward with this as is, and if you want to continue to pursue this @otaviof , we start with a community discussion in the Monday meeting, and see where that goes from there

**Review by gabemontero — COMMENTED** [review]
so @coreydaley do you plan to add any valueFrom secret/configmap related options on the build/buildrun creates you added ... i.e. a --valueFromSecret=<secret name> sort of thing

File: test/e2e/e2e.bats
Comment (otaviof) [reply test/e2e/e2e.bats]: I think vendoring the external dependency is easier due being part of regular commits, we include the `vendor` directory in the project itself, as usual, and it makes everything simpler IMO. Git submodule is something I only have bad memories, and compared to the `vendor` option, it does require extra steps.

**Comment by coreydaley** [comment]
> so @coreydaley do you plan to add any valueFrom secret/configmap related options on the build/buildrun creates you added ... i.e. a --valueFromSecret= sort of thing

No, @adambkaplan and I discussed it yesterday. I also don't think that `oc` ever added that capability (probably due to complexity)

File: pkg/shp/cmd/buildrun/create.go
Comment (otaviof) [reply pkg/shp/cmd/buildrun/create.go:70]: Sure! Make sense, @gabemontero!

File: test/e2e/e2e.bats
Comment (coreydaley) [reply test/e2e/e2e.bats]: Git sub-modules are really the de-facto way to include external libraries like this in your repository. 
Vendoring a project that has absolutely no Go code seems pretty sketchy.

**Comment by gabemontero** [comment]
> > so @coreydaley do you plan to add any valueFrom secret/configmap related options on the build/buildrun creates you added ... i.e. a --valueFromSecret= sort of thing
> 
> No, @adambkaplan and I discussed it yesterday. I also don't think that `oc` ever added that capability (probably due to complexity)

I'm fine waiting and seeing if users ask for it ... as noted OCP / oc users either never did, or not enough that we bothered to add it

thanks

File: test/e2e/e2e.bats
Comment (otaviof) [reply test/e2e/e2e.bats]: Then, I think we can benefit from a more in-depth discussion about the git-submodules, @coreydaley. Seems a bit too much given we need a couple of assertions we can easily achieve with shell-script.

File: pkg/shp/cmd/buildrun/create.go
Comment (gabemontero) [reply pkg/shp/cmd/buildrun/create.go:70]: cool thanks !

File: test/e2e/e2e.bats
Comment (gabemontero) [reply test/e2e/e2e.bats]: sounds like another topic for Monday / community meeting

**Comment by gabemontero** [comment]
/assign @gabemontero 
/assign @otaviof 
/assign @adambkaplan 

@adambkaplan - 
aside from vendoring in @coreydaley 's https://github.com/shipwright-io/build/pull/817 changes once they merge, the main outstanding item here before merging this for a v.0.6.0 release cut is making a final decision on how we pull in for reuse existing functions from the bats repo into this repo 

the discussion thread starts with https://github.com/shipwright-io/cli/pull/27#discussion_r721156689

currently, there are 2 votes for git modules and one for I think use of go modules ... or minimally a lack of being convinced git submodules is the way to go

@coreydaley 's e2es need to be reworked if we move off of git submodules

to date I have not been ready to say majority wins even if we don't reach 100% consensus, and fully tag / unblock this, given @otaviof 's existing questions on what @coreydaley and I agree on

but if we want to include this as part of cutting a v0.6.0 release this week, a possible "no 100% consensus exists but we are moving forward" decision may be needed

so please catch up if you can on the discussion thread and chime in ... thanks

**Comment by coreydaley** [comment]
The tests for this new feature will not pass until a new release is cut for shipwright-io/build and this repo is updated to use that new release.

**Comment by adambkaplan** [comment]
Regarding BATS - I wonder if there is a third option, which is to use the official container image to run bats tests. It is not clear based on the docs if the container image includes the supporting libraries (like assert).

https://bats-core.readthedocs.io/en/stable/installation.html#running-bats-in-docker

**Comment by coreydaley** [comment]
> Regarding BATS - I wonder if there is a third option, which is to use the official container image to run bats tests. It is not clear based on the docs if the container image includes the supporting libraries (like assert).
> 
> https://bats-core.readthedocs.io/en/stable/installation.html#running-bats-in-docker
I don't think that really helps users who want to run the tests locally, or run them during development.

File: go.mod
Comment (gabemontero) [inline go.mod:8]: you said in slack you were going to bump this to the newly minted v0.6.0, right @coreydaley ?

Comment (gabemontero) [reply go.mod:8]: ah .. I do see https://github.com/shipwright-io/cli/pull/51though that only touches the Makefile/install script

Comment (coreydaley) [reply go.mod]: good catch, updated.

File: docs/testing.md
Comment (otaviof) [inline docs/testing.md:62]: This block should follow the formatting of the rest of the markdowns here, as in:

```sh
# Option 1: Clone the repository and populate submodules right away
git clone --recurse-submodules --depth 1 https://github.com/shipwright-io/cli.git

# Option 2: Populate submodules after cloning
cd /your/project/directory/cli
git submodule init
git submodule update
```

File: hack/run-e2e.sh
Comment (otaviof) [inline hack/run-e2e.sh:4]: nit: Here we could use `/test/e2e/bats/core/bin/bats --recursive test/e2e/*.bats` instead.

File: test/e2e/e2e.bats
Comment (otaviof) [inline test/e2e/e2e.bats:31]: There's a typo on the `--buildref-name` curly brace position. It should look like the following:

```bash
run shp buildrun create ${buildrun_name} --buildref-name=${build_name}
```

File: test/e2e/e2e.bats
Comment (otaviof) [inline test/e2e/e2e.bats:3]: In `helpers.sh` we define a function named `fail` which [collides with bats-core](https://github.com/bats-core/bats-core/blob/8e9f35aa794e38cef6cc9da21003730cce00c3bb/libexec/bats-core/bats-format-pretty#L87). I think the best course of action is to rename it, or maybe removing it completely.

Comment (coreydaley) [reply test/e2e/e2e.bats:3]: updated in current push

File: docs/testing.md
Comment (coreydaley) [reply docs/testing.md:63]: it should be fixed now.

**Pull request merged**
